### PR TITLE
Please review: Memory Object handling

### PR DESCRIPTION
**Pull request by penguineer**
Please review:
This PR contains an API for pointer ownership handling and an integration into the existing configuration management.

The changes aim towards #9, the remaining features will be added after a review.

### CODE REVIEW

**Comment by penguineer** [comment]
The memory object could even be extended with a free-function pointer to specify a function that should be called to free the resources of the memory object.

File: mosqagent.h
Comment (LeSpocky) [inline mosqagent.h:28]: If the API user should not manipulate this directly, you can hide this struct from the public header and just include a forward declaration in the public header. Declare the struct in the c file or a private header instead. This way you can ensure the caller can not easily manipulate the internal members, and you can change the internals later without breaking API user code.

File: mosqagent.h
Comment (LeSpocky) [inline mosqagent.h:33]: If you don't change the contents of `*mo` you should add `const` to the parameter like this:
`void* mqtta_mo_ptr(const struct mqtta_memory_object *mo);`

File: mosqagent.h
Comment (LeSpocky) [inline mosqagent.h:38]: Should be const parameter like with `mqtta_mo_ptr()`. More important: The documentation what ownership means, is missing. If this returns `true`, who is owning the thing then?

File: mosqagent.h
Comment (LeSpocky) [inline mosqagent.h:50]: You may add `\warning` or `\note` or something which stands at least a little out here for Doxygen documentation.

File: mosqagent.h
Comment (LeSpocky) [inline mosqagent.h:167]: What about using the Doxygen command `\note` here?

File: mqtt-clock.c
Comment (LeSpocky) [inline mqtt-clock.c:186]: `mqtta_mo_ptr()` returns the pointer stored in the memory object. You didn't set a pointer and you ignore the result. This call makes no sense.

**Review by LeSpocky — CHANGES_REQUESTED** [review]
Okay in general, you may have another opinion on my remarks to the first changeset of the series and keep it as is instead of changing it before merging, this should be no stopper.

However I don't approve the PR because of the senseless call in the last changeset.

Everything else, especially `const` where possible and documentation improvements would be nice to have, but not critical.

File: mosqagent.h
Comment (penguineer) [reply mosqagent.h:28]: I thought about this: If I hide the struct, only pointers can be used. Which means there are heap objects of structs with meta-information about heap objects (and again the question who has to clean those up). 

My idea was that this struct is embedded directly. The reason for using pointers in the API is that copies of this struct should not be necessary and the user can decide where the struct resides.

**Comment by penguineer** [comment]
Apart from the above improvements, I will make one major change:
The ownership boolean will be replaced by a function pointer to a de-allocator function (signature compatible to _free_). If the function pointer is null, the object will be left alone, otherwise the de-allocator will be called. 
With this change, the basic function is kept: User can decide who should clean up the memory object. Additionally the correct method can be chosen when the memory object is created, leading to more flexibility in its use, e.g. the configuration struct can have a de-allocator which also frees the strings inside the struct.

File: mqtt-clock.c
Comment (penguineer) [reply mqtt-clock.c:186]: Yes, these are left-overs from some testing and will be evicted.

File: mosqagent.h
Comment (penguineer) [reply mosqagent.h:50]: Wasn't aware of these.

http://www.stack.nl/~dimitri/doxygen/manual/commands.html

File: mosqagent.h
Comment (LeSpocky) [reply mosqagent.h:28]: You are right, and you use it like this in mqtt-clock.c here:

```
    struct mqtta_memory_object mo;
    mqtta_mo_ptr(&mo);
```

I overlooked this part. If you use it like this, you can not hide the struct.

Maybe this is some question of taste. I saw other libraries where you request some kind of handle and the allocation of the struct behind happens dynamically inside the lib.

Comment (penguineer) [reply mosqagent.h:28]: Yes… I am not a fan of handle-based systems, though.This would mean replacing one handle (pointer) by a far less powerful handle just so that the user cannot manipulate one value.

On the other hand: In those cases where the users hand over a pointer, they need to specify the da-allocator anyways. So even with a handle-based system there is so much margin for error that I do not think it is worth the effort.

After all, I wanted to make (my) life a bit easier.

Comment (penguineer) [reply mosqagent.h:28]: There has been much discussion about this topic regarding smart pointers in C++ standardization.

Comment (LeSpocky) [reply mosqagent.h:28]: Maybe I was just irritated by this comment:
```
/**
 * \brief A memory object that stores ownership information.
 *
 * Never manipulate the values directly! Always use mqtta_mo_move and
 * mqtta_mo_set to change pointers. Ownership information is managed
 * automatically by these functions.
 */
```
So if this library exposes the memory object, I'm fine with that, as long as the intended use is documented.

**Comment by penguineer** [comment]
Due to connectivity problems with Vodafone delivery of an updates branch is delayed. …

**Pull request closed without merging**
